Core/HLE/sceNetAdhoc.cpp

```diff
@@ -113,8 +120,457 @@ void __NetAdhocShutdown() {
+	if (ret >= 0 && ret <= *req.length) {
+		ret = recvfrom(uid, (char*)req.buffer, *req.length, 0, (sockaddr*)&sin, &sinlen);
+		// UDP can also receives 0 data, while on TCP receiving 0 data = connection gracefully closed, but not sure whether PDP can send/recv 0 data or not tho
+		if (ret > 0) {
+			DEBUG_LOG(SCENET, "sceNetAdhocPdpRecv[%i:%u]: Received %u bytes from %s:%u\n", req.id, getLocalPort(uid), ret, inet_ntoa(sin.sin_addr), ntohs(sin.sin_port));
+
+			// Peer MAC
+			SceNetEtherAddr mac;
+
+			// Find Peer MAC
+			if (resolveIP(sin.sin_addr.s_addr, &mac)) {
+				// Provide Sender Information
+				*req.remoteMAC = mac;
+				*req.remotePort = ntohs(sin.sin_port) - portOffset;
+
+				// Save Length
+				*req.length = ret;
+
+				// Update last recv timestamp
+				peerlock.lock();
+				auto peer = findFriend(&mac);
+				if (peer != NULL) peer->last_recv = CoreTiming::GetGlobalTimeUsScaled();
+				peerlock.unlock();
+			}
+			// Unknown Peer
+			else {
+				*req.length = ret;
+				*req.remotePort = ntohs(sin.sin_port) - portOffset;
+
+				WARN_LOG(SCENET, "sceNetAdhocPdpRecv[%i:%u]: Received %i bytes from Unknown Peer %s:%u", req.id, getLocalPort(uid), ret, inet_ntoa(sin.sin_addr), ntohs(sin.sin_port));
+			}
+		}
+		result = 0;
+	}
+	// On Windows: recvfrom on UDP can get error WSAECONNRESET when previous sendto's destination is unreachable (or destination port is not bound yet), may need to disable SIO_UDP_CONNRESET error
+	else if (sockerr == EAGAIN || sockerr == EWOULDBLOCK || sockerr == ECONNRESET || sockerr == ETIMEDOUT) {
+		u64 now = (u64)(real_time_now() * 1000000.0);
+		if (req.timeout == 0 || now - req.startTime <= req.timeout) {
+			
+			return -1;
+		}
+		else
+			result = ERROR_NET_ADHOC_TIMEOUT;
+	}
+	else
+		result = ERROR_NET_ADHOC_INVALID_ARG; // ERROR_NET_ADHOC_DISCONNECTED
+
+	if (ret == SOCKET_ERROR)
+		DEBUG_LOG(SCENET, "sceNetAdhocPdpRecv[%i]: Socket Error (%i)", req.id, sockerr);
+
+	return 0;
+}
+
+int DoBlockingPdpSend(int uid, AdhocSocketRequest& req, s64& result, AdhocSendTargets& targetPeers) {
+	SceNetAdhocPdpStat* pdpsocket = pdp[req.id - 1];
+
+	result = 0;
+	bool retry = false;
+	for (auto peer = targetPeers.peers.begin(); peer != targetPeers.peers.end(); ) {
+		// Fill in Target Structure
+		sockaddr_in target;
+		target.sin_family = AF_INET;
+		target.sin_addr.s_addr = peer->ip;
+		target.sin_port = htons(peer->port + ((isOriPort && !isPrivateIP(peer->ip)) ? 0 : portOffset));
+
+		int ret = sendto(pdpsocket->id, (const char*)req.buffer, targetPeers.length, 0, (sockaddr*)&target, sizeof(target));
+		int sockerr = errno;
+
+		if (ret >= 0) {
+			DEBUG_LOG(SCENET, "sceNetAdhocPdpSend[%i:%u](B): Sent %u bytes to %s:%u\n", uid, getLocalPort(pdpsocket->id), ret, inet_ntoa(target.sin_addr), ntohs(target.sin_port));
+			// Remove successfully sent to peer to prevent sending the same data again during a retry
+			peer = targetPeers.peers.erase(peer);
```

Comment:
I'm a bit worried with this `erase`, i wanted to erase the content of the global map, but not sure whether this code will affect globally or this `peers` is only a copy, i only want to retry sending on failed peers and not resending to the one already successful on the next rescheduled event

Reply:
`targetPeers` is a reference, so any changes to `targetPeers.peers` will affect that reference.

`foo.erase()` mutates foo, you'd need to do `std::deque<AdhocSendTarget>(foo).erase` or something to make it use a copy, and even then `peer` would not be a valid iterator to the copy, I assume.

-[Unknown]